web/src/bootstrap_typeahead.ts

```diff
@@ -303,52 +311,57 @@ class Typeahead<ItemType extends string | object> {
     }
 
     show(): this {
+        if (this.shown) {
+            return this;
+        }
+
+        // Call this early to avoid duplicate calls.
+        this.shown = true;
+
         const header_text_html = this.header_html();
         if (header_text_html) {
             this.$header.find("span#typeahead-header-text").html(header_text_html);
             this.$header.show();
         } else {
             this.$header.hide();
         }
+        this.mouse_moved_since_typeahead = false;
 
-        // If a parent element was specified, we shouldn't manually
-        // position the element, since it's already in the right place.
-        if (this.parentElement === undefined) {
-            let pos;
-
-            pos = this.input_element.$element[0].getBoundingClientRect();
-
-            pos = $.extend({}, pos, {
-                height: this.input_element.$element[0].offsetHeight,
-                // Zulip patch: Workaround for iOS safari problems
-                top: this.input_element.$element.get_offset_to_window().top,
-            });
-
-            let top_pos = pos.top + pos.height;
-            if (this.dropup) {
-                top_pos = pos.top - this.$container.outerHeight()!;
-            }
-
-            // Zulip patch: Avoid typeahead going off top of screen.
-            if (top_pos < 0) {
-                top_pos = 0;
-            }
-
-            this.$container.css({
-                top: top_pos,
-                left: pos.left,
-            });
+        if (this.parentElement) {
+            this.$container.show();
+            // We don't need tippy to position typeaheads which already know where they should be.
+            return this;
         }
+        this.instance = tippy(this.input_element.$element[0], {
+            maxWidth: "none",
+            theme: "popover-menu",
+            placement: this.dropup ? "top-start" : "bottom-start",
+            interactive: true,
+            appendTo: () => document.body,
+            showOnCreate: true,
+            content: this.$container[0],
+            trigger: "manual",
+            arrow: false,
+            offset: [0, 0],
+            hideOnClick: false,
+            onHidden: () => {
+                assert(this.instance !== undefined);
+                this.instance.destroy();
+                this.instance = undefined;
+            },
+        });
```

Comment:
Can you add a batch of comments documenting the reasons for some of the more important choices here?

Reply:
Done